core/src/main/scala/org/typelevel/paiges/Doc.scala

```diff
@@ -537,13 +544,25 @@ object Doc {
    */
   private[paiges] case object Empty extends Doc
 
+  /**
+   * Render 'default' except when flattened.
+   * Invariant: width(default) <= width(whenFlat)
+   * Invariant: default != whenFlat
+   */
+  private[paiges] case class FlatAlt(default: Doc, whenFlat: Doc) extends Doc
+
+  /**
+   * Used during rendering to indicate a hard line break which
+   * can't be flattened.
+   * Invariant: Fail only appears on the left-hand side (i.e. the flattened side) of a Union
+   */
+  private[paiges] case object Fail extends Doc
```

Comment:
can't this be returned by `def flatten`? We should have a test for this invariant:

```scala
forAll { a: Doc =>
  // something like this except a recursive check:
  a match {
    case Union(Fail, _) => assert(true)
    case Fail => assert(false)
    case _ => assert(true)
  }

  // more over:
  assert(a.flatten != Fail)
}
```

Can you describe how this can be rendered? And what it maxWidth, etc... is?

We should have stronger tests if we are adding a new variant.

Reply:
Yes, `Fail` is frequently returned by `flatten`, as it is the instruction to `best` to ignore the current path, and use the next one.   

I'm trying to figure out what the laws should be now.  We don't need to add `Fail` randomly to the generators, since it's only inserted by `Flatten`.  `FlatAlt` currently only comes from `line`, etc, so I'll throw those in.  But indeed, we need to clarify in the laws where `Fail` is allowed.  I'll submit another patch shortly.